src/chainparamsbase.h

```diff
@@ -24,11 +24,14 @@ class CBaseChainParams
     const std::string& DataDir() const { return strDataDir; }
     int RPCPort() const { return nRPCPort; }
 
-protected:
-    CBaseChainParams() {}
+    CBaseChainParams(const std::string& data_dir, int rpc_port) : nRPCPort(rpc_port), strDataDir(data_dir) {}
 
+private:
     int nRPCPort;
     std::string strDataDir;
+
+    /** Default constructor is private to force the use of the parametrized one */
+    CBaseChainParams() {}
```

Comment:
Having a non-default constructor already prevents automatic declaration of the default constructor; but "CBaseChainParams() = delete;" seems clearer still.

Also, since nRPCPort and strDataDir are set via initializer list, and aren't fiddled with later (they're private and all the member functions are marked const), they could be marked "const" too.

Tested with these changes:

```
+    const int nRPCPort;
+    const std::string strDataDir; 

+    CBaseChainParams() = delete;
+    CBaseChainParams(const CBaseChainParams&) = delete;
+    CBaseChainParams& operator=(const CBaseChainParams&) = delete;
```

Reply:
> Having a non-default constructor already prevents automatic declaration of the default constructor

Does it? I really think that's not the case, do you have any reference?

> but "CBaseChainParams() = delete;" seems clearer still.

Perhaps. I've never seen this contruct but if people prefer this I'm happy to change it.

Reply:
(`= delete` is new in C++11)

Reference: http://en.cppreference.com/w/cpp/language/default_constructor

> If no user-declared constructors of any kind are provided for a class type (struct, class, or union), the compiler will always declare a default constructor as an inline public member of its class. 
>
> [[c++11:]] If some user-declared  constructors are present, the user may still force the automatic  generation of a default constructor by the compiler that would be  implicitly-declared otherwise with the keyword default.

Example:
```
class X { public: int a; char b; X(int _a, char _b) : a(_a), b(_b) { } };
X x;
```

gcc:
```
test.cc:2:3: error: no matching function for call to ‘X::X()’
 X x;
   ^
test.cc:1:34: note: candidate: X::X(int, char)
test.cc:1:34: note:   candidate expects 2 arguments, 0 provided
test.cc:1:7: note: candidate: constexpr X::X(const X&)
test.cc:1:7: note:   candidate expects 1 argument, 0 provided
test.cc:1:7: note: candidate: constexpr X::X(X&&)
test.cc:1:7: note:   candidate expects 1 argument, 0 provided
```

clang:
```
test.cc:2:3: error: no matching constructor for initialization of 'X'
X x;
  ^
test.cc:1:7: note: candidate constructor (the implicit copy constructor) not
      viable: requires 1 argument, but 0 were provided
test.cc:1:7: note: candidate constructor (the implicit move constructor) not
      viable: requires 1 argument, but 0 were provided
test.cc:1:34: note: candidate constructor not viable: requires 2 arguments,
      but 0 were provided
class X { public: int a; char b; X(int _a, char _b) : a(_a), b(_b) { } };
                                 ^
```

Reply:
Thinking more about this, I see how your interest in also prohibiting the copy constructor and the override of the assign operator may make sense in many cases, but I think in this case we're just fine prohibiting only the default contructor. Anything copied from an instance initialized with the parametrized contructor should be fine.
But thank you for teaching me newer c++ features, I love to be able to ditch that hackish design pattern.
I will ```
CBaseChainParams() = delete;
```
Actually the fact that an upgrade of the language supports what I was "simulating" gives me much more confidence on that part. But thank you again for stopping me from reinventing the wheel.